kernel/src/table_changes/data_read.rs

```diff
@@ -0,0 +1,217 @@
+use std::iter;
+
+use itertools::Either;
+use roaring::RoaringTreemap;
+use url::Url;
+
+use super::scan_file::{CdfScanFileType, ResolvedCdfScanFile, UnresolvedCdfScanFile};
+use crate::actions::deletion_vector::selection_treemap_to_bools;
+use crate::table_changes::scan_file::CdfScanFile;
+use crate::{DeltaResult, Engine, Error};
+
+/// Resolves the deletion vectors for an [`UnresolvedCdfScanFile`]. This function handles two
+/// types of `CdfScanFile`s:
+/// 1. The first case is a [`CdfScanFile`] paired with a remove deletion vector. The `scan_type`
+///    must be [`CdfScanFileType::Add`]. In this case, both the add and remove deletion vectors are
+///    read if they exist. Then, we find the set of rows in the scan file that have been added, and
+///    the set of rows that have been removed. The set of removed rows (if any) will be represented
+///    by a [`ResolvedCdfScanFile`] with `scan_type` = [`CdfScanFileType::Remove`]. The set of
+///    added rows (if any) will be represented by a [`ResolvedCdfScanFile`] with `scan_type` =
+///    [`CdfScanFileType::Add`].
+///
+///    Note: We allow the possibility for there to be both added rows, and removed rows for a
+///    single add/remove pair.
+/// 2. The second case handles all other add, remove, and cdc [`CdfScanFile`]s. These will simply
+///    have the deletion vector read (if present), and each is converted into a [`ResolvedCdfScanFile`].
+///    No changes are made to the `scan_type`.
+#[allow(unused)]
+pub(crate) fn resolve_scan_file_dv(
+    engine: &dyn Engine,
+    table_root: &Url,
+    scan_file: UnresolvedCdfScanFile,
+) -> DeltaResult<impl Iterator<Item = ResolvedCdfScanFile>> {
+    let UnresolvedCdfScanFile {
+        scan_file,
+        remove_dvs,
+    } = scan_file;
+    let paired_rm_dv = remove_dvs.get(&scan_file.path);
+    match (&scan_file.scan_type, paired_rm_dv) {
+        (CdfScanFileType::Add, Some(rm_dv)) => {
+            // Helper function to convert a treemap to a [`ResolvedCdfScanFile`]. The `scan_type`
+            // of the [`ResolvedCdfScanFile`] is set to `out_type` This returns an empty iterator
+            // if nothing is selected.
+            fn treemap_to_iter(
+                selection_treemap: RoaringTreemap,
+                mut scan_file: CdfScanFile,
+                out_type: CdfScanFileType,
+            ) -> impl Iterator<Item = ResolvedCdfScanFile> {
+                if selection_treemap.is_empty() {
+                    // Nothing has been selected, we do not read this data file
+                    Either::Left(iter::empty())
```

Comment:
Problem: The DV for a file can be shorter than the file's row count, with missing entries treated as if they were `true` (= keep). See doc comments for [ScanResult::full_mask](https://github.com/delta-io/delta-kernel-rs/blob/main/kernel/src/scan/mod.rs#L150). My tired Friday pm brain is struggling to grok whether computing a symmetric difference for a remove DV breaks that assumption? But it's _definitely_ not safe to skip file just because its raw physical DV is empty (current code in PR), or even just because all DV entries seem to be false (because every DV has an implied padding of true entries past-end, out to the data length which we won't know until we have an `EngineData` to work with).

Reply:
Hmm I may be skipping something, but here's what I know:
- This skipping here only happens for add/remove pairs of deletion vectors.
-  the treemap here represents the *selected* rows based on the dv. So far in our code we've treated treemaps exclusively as deletion bitmaps. Our dv resolving results in us generating a **selection** treemap.  Hence if the selection treemap no bit set, everything would be false => not selected/skipped

I still think should be safe to skip?

Reply:
I've added a comment explaining why the treemap is [treated as a selection vector ](https://github.com/delta-io/delta-kernel-rs/pull/568/files#diff-6f030aa38958bdafebc42886e31c46454447d689b49e0b5f1291d37a6645fdcfR71).

Reply:
I think the issue is, let's say your data is:

`['a', 'b', 'c']` and your DV is:
`[0] // row 0 is deleted`

the _implied_ DV is `[false, true, true]`.

This matters when you're just using DVs "as is". _However_, here you're only look for rows that _changed_, so if there are a bunch of indices not mentioned in either treemap, they did not change, so should not be returned here. 

I think your comment below should work through with the indices first, since that's what the treemap has, and the set difference operations to show the resulting indicies, and _then_ the conversion to boolean selection vectors, to ensure things work as we expect.

Re it being _definitely_ wrong if the vector is empty, that's certainly true for a standard scan, but I think here it's potentially correct if it means that nothing _changed_ between the two vectors.

Reply:
Maybe to add on. All the important operations here occur where things are just treemaps. Only once we've settled on the final indicies which have changed do we convert to a boolean vector, which is the same as in our scan code. But, because we're using _selection_ vectors here, we don't need to worry about extending, because the implied extra stuff is masked out.

We should probably put a bit `// ===== IMPORTANT =====` comment somewhere that calls out that a CDF scan uses a _selection_ vector while a "normal" scan uses a _deletion_ one.

Reply:
I hopefully explained it more thoroughly this time. Added some notation to be clear at each point what type of treemap we're working with. Also split up selection vector generation into its own section.